### PR TITLE
Column Compare Rules

### PR DESCRIPTION
**Pull request by Snotpus**
I don't necessarily want you to try and accept this PR... but I wanted to created it so you could see what kinds of things I was attempting to do. It is bad.

### CODE REVIEW

File: rumydata/field.py [hunk 1]
```diff
@@ -116,7 +134,7 @@ def _comparison_columns(self) -> set:
         compares = set()
         for r in self.rules:
             if issubclass(type(r), clr.ColumnComparisonRule):
-                compares.add(r.compare_to)
+                compares.add(r.compare_to) if isinstance(r.compare_to, str) else [compares.add(x) for x in r.compare_to]
```
Comment (Mikuana) [inline rumydata/field.py:137]: Ooh that's a really nice use of an inline add. I didn't realize you could do it like that.

File: tests/test_rules_cell.py [hunk 2]
```diff
@@ -375,3 +375,39 @@ def test_date_lt(comparison: str, value: str, expected: bool):
 def test_greater_than_column(compared: str, data, expected: bool):
     r = GreaterThanColumn(compared)
     assert r._evaluator()(*r._prepare(data)) is expected
+
```
Comment (Snotpus) [inline tests/test_rules_cell.py:378]: The tests still need work... I didn't go back and clean them up or really work on adding any additional tests just due to (my perceived) available time to finish...

File: rumydata/rules/cell.py [hunk 3]
```diff
@@ -546,14 +546,30 @@ class ColumnComparisonRule(Rule):
 
     _default_args = ('x',)
 
-    def __init__(self, compare_to: str):
+    def __init__(self, compare_to: [str, List]):
```
Comment (Snotpus) [inline rumydata/rules/cell.py:549]: change type hint to `Union[str, List[str]]`

File: rumydata/rules/cell.py [hunk 4]
```diff
@@ -562,3 +578,45 @@ def _evaluator(self):
...
+
+    def __init__(self, compare_to: [str, List]):
+        super().__init__(compare_to=compare_to)
+
+    def _prepare(self, data: Tuple[str, Dict]) -> tuple:
+        return data
+
+    def null_ok(self, data):
+        empty_val = data[0] in ['', False]
+        empty_compare = True
```
Comment (Snotpus) [inline rumydata/rules/cell.py:603]: Walrus operator? ;)

File: rumydata/rules/cell.py [hunk 5]
```diff
@@ -562,3 +578,45 @@ def _evaluator(self):
...
+
+class NotNullIfCompare(ColumnComparisonRule):
+
+    def __init__(self, compare_to: [str, List]):
+        super().__init__(compare_to=compare_to)
+
+    def _prepare(self, data: Tuple[str, Dict]) -> tuple:
+        return data
+
+    def null_ok(self, data):
```
Comment (Snotpus) [inline rumydata/rules/cell.py:601]: rename _null_ok()

**Comment by Mikuana** [comment]
I added a `dev` branch for merging these new features. I added `tests.test_other.test_complex_good`, which is intended to be a test of the full complexity of the package. I think it would be helpful if you added some data and layout configurations to test your conditional null additions.

File: tests/test_other.py [hunk 6]
```diff
@@ -288,23 +288,54 @@ def test_empty_column_bad(basic, basic_good_with_empty):
 
 
 @pytest.fixture()
```
Comment (Snotpus) [inline tests/test_other.py:290]: Tried to follow your lead with the `test_complex_good`... I extended it to include specific tests for the kinda weird 'conditional' cell rules (`NotNullIfCompare`, `OtherCantExist`, and `OtherMustExist`)... I then attempted to replicate the approach to ensure that a bad complex file will trip when expected... I feel the way I extended the test makes them a bit hard to parse since it's trying to check several things at once, but opted to stay the course since we are trying to test the combination of these things aferall!

**Comment by Mikuana** [comment]
@Snotpus this looks good to me, but I'm going to try playing with the GitHub actions to run the tests on Windows so that we can try to recreate the failing tests that you saw on your machine.

**Comment by Snotpus** [comment]
I'm not really sure what to do with this info yet, but the failing teardowns related to Excel seem to stem somehow from using the `read_only` mode when opening the workbeen via openpyxl. See https://stackoverflow.com/questions/31416842/openpyxl-does-not-close-excel-workbook-in-read-only-mode (ignore the accepted answer, further comments talk specifically about the `read_only` mode). I ways able to quickly try removing `read_only=True` and the tests pass as expected! I'm not sure that we want to rethink using read_only mode or not

**Comment by Mikuana** [comment]
@Snotpus I've modified the `dev` branch GitHub actions script to test on Ubuntu, Mac, and Windows. The Windows test fails, the same way that it does on your workstation. So good job finding a bug! Let me know once you've got a solution to the Windows error.

**Comment by Mikuana** [comment]
We could try modifying the code for the ExcelFile reader to incorporate the method on [this](https://stackoverflow.com/a/43789779/1958969) SO answer.

```
def __enter__(self) -> Iterable:
    from openpyxl import load_workbook

    with open(file, "rb") as f:
        in_mem_file = io.BytesIO(f.read())
    wb = load_workbook(in_mem_file, read_only=True, data_only=True)
    sheet_name = self.excel_kwargs.get('sheet')
    ws = wb[sheet_name] if sheet_name else wb.active
    return ws.values

```

**Comment by Mikuana** [comment]
> We could try modifying the code for the ExcelFile reader to incorporate the method on [this](https://stackoverflow.com/a/43789779/1958969) SO answer.
> 
> ```
> def __enter__(self) -> Iterable:
>     from openpyxl import load_workbook
> 
>     with open(file, "rb") as f:
>         in_mem_file = io.BytesIO(f.read())
>     wb = load_workbook(in_mem_file, read_only=True, data_only=True)
>     sheet_name = self.excel_kwargs.get('sheet')
>     ws = wb[sheet_name] if sheet_name else wb.active
>     return ws.values
> ```

Maybe not. I tried this in another branch and it didn't work.

**Comment by Mikuana** [comment]
@Snotpus the Windows test fail is related to the `openpyxl` issue that you found on SO. There's an open PR to address this, which we can track in issue #58. I suggest we make the windows OS tests non-required until this issue is resolved, and move forward with merging your branch.

What do you think? Ready to give this the green light?

**Comment by Snotpus** [comment]
nice job finding that openpxl issue... I had tried looking for something but didn't spend enough time looking apparently! I think I'm good with merging this stuff in if you are :)

**Comment by Mikuana** [comment]
@Snotpus no worries; the PR was only opened 5 hours ago, so I think we're just insanely luck (or unlucky) with our timing in discovering this issue.
